File: Classes/DrawUIBezierPathFitToUIView.m
//
//  DrawUIBezierPathFitToUIView.m
//
//  Created by Mossuru777 on 2014/11/14.
//  Copyright (c) 2014 Mossan. All rights reserved.
//

#import "DrawUIBezierPathFitToUIView.h"

@implementation DrawUIBezierPathFitToUIView
{
	@private
	CGSize viewSize;
	NSMutableArray *paths;
}

- (id)init
{
	[[NSException exceptionWithName:@"NoArgumentsInitializingException"
                             reason:@"Can't initialize with init method. Use initWithView method."
                           userInfo:nil
     ] raise];
	return nil;
}

- (id)initWithView:(UIView *)view
{
	self = [super init];
	if(self)
	{
		if(view == nil)
		{
			[[NSException exceptionWithName:@"NilObjectException"
                                     reason:@"Can't initialize UIView with nil."
                                   userInfo:nil
             ] raise];
		}
		viewSize = view.frame.size;
		paths = [[NSMutableArray alloc] init];
		self.scale = 1.0;
	}
	return self;
}

- (void)addPath:(UIBezierPath *)path fillColor:(UIColor *)fillColor strokeColor:(UIColor *)strokeColor
{
	// check exist path in array
	for (NSArray *path_info in paths)
	{
		if([path_info[0] isEqual:path])
		{
			return;
		}
	}

	// add path to array
	if(fillColor != nil)
	{
		if(strokeColor != nil)
		{
			[paths addObject:[[NSArray alloc] initWithObjects:path, fillColor, strokeColor, nil]];
		}
		else {
			[paths addObject:[[NSArray alloc] initWithObjects:path, fillColor, [NSNull null], nil]];
		}
	}
	else {
		if(strokeColor != nil)
		{
			[paths addObject:[[NSArray alloc] initWithObjects:path, [NSNull null], strokeColor, nil]];
		}
		else {
			[NSException raise:@"NoDrawPathException" format:@"Arguments indicate NO draw path!!!"];
		}
	}
}

- (void)drawScaledPathFitToView
{
#if __LP64__
	CGRect totalRect = CGRectMake(DBL_MAX, DBL_MAX, 0, 0);
#else
	CGRect totalRect = CGRectMake(FLT_MAX, FLT_MAX, 0, 0);
#endif
	CGFloat fit_scale;
	CGFloat centering_margin_x = 0;
	CGFloat centering_margin_y = 0;
	CGFloat move_x;
	CGFloat move_y;

	// calculate total bounds / origin
	for (NSArray *path_info in paths)
	{
		UIBezierPath *path = path_info[0];
		NSObject *fillColor = path_info[1];
		NSObject *strokeColor = path_info[2];

		// reset path scaling
		[path applyTransform:CGAffineTransformIdentity];

		// extends bounds / origin
		if([strokeColor isKindOfClass:[UIColor class]])
		{
			// width
			if(path.bounds.origin.x + path.bounds.size.width + (path.lineWidth / 2) > totalRect.size.width)
			{
				totalRect.size.width = path.bounds.origin.x + path.bounds.size.width + (path.lineWidth / 2);
			}

			// x origin
			if(path.bounds.origin.x - (path.lineWidth / 2) < totalRect.origin.x)
			{
				totalRect.origin.x = path.bounds.origin.x - (path.lineWidth / 2);
			}

			// height
			if(path.bounds.origin.y + path.bounds.size.height + (path.lineWidth / 2) > totalRect.size.height)
			{
				totalRect.size.height = path.bounds.origin.y + path.bounds.size.height + (path.lineWidth / 2);
			}

			// y origin
			if(path.bounds.origin.y - (path.lineWidth / 2) < totalRect.origin.y)
			{
				totalRect.origin.y = path.bounds.origin.y - (path.lineWidth / 2);
			}
		}
		else if([fillColor isKindOfClass:[UIColor class]])
		{
			// width
			if(path.bounds.origin.x + path.bounds.size.width > totalRect.size.width)
			{
				totalRect.size.width = path.bounds.origin.x + path.bounds.size.width;
			}

			// x origin
			if(path.bounds.origin.x < totalRect.origin.x)
			{
				totalRect.origin.x = path.bounds.origin.x;
			}

			// height
			if(path.bounds.origin.y + path.bounds.size.height > totalRect.size.height)
			{
				totalRect.size.height = path.bounds.origin.y + path.bounds.size.height;
			}

			// y origin
			if(path.bounds.origin.y < totalRect.origin.y)
			{
				totalRect.origin.y = path.bounds.origin.y;
			}
		}
	}

	// calculate scale / centering margin / move
	if(self.scale > 1.0 || self.scale <= 0)
	{
		self.scale = 1.0;
	}
	if(viewSize.width / totalRect.size.width < viewSize.height / totalRect.size.height)
	{
		fit_scale = viewSize.width / totalRect.size.width * self.scale;
	}
	else {
		fit_scale = viewSize.height / totalRect.size.height * self.scale;
	}
	centering_margin_x = (viewSize.width / 2) - (totalRect.size.width * fit_scale / 2);
	centering_margin_y = (viewSize.height / 2) - (totalRect.size.height * fit_scale / 2);
	move_x = totalRect.origin.x * fit_scale;
	move_y = totalRect.origin.y * fit_scale;

	// draw scaled path
	for(NSArray *path_info in paths)
	{
		UIBezierPath *path = path_info[0];
		NSObject *fillColor = path_info[1];
		NSObject *strokeColor = path_info[2];

		// scaling path
		path.lineWidth = path.lineWidth * fit_scale;
		[path applyTransform:CGAffineTransformMakeScale(fit_scale, fit_scale)];

		// move path
		[path applyTransform:CGAffineTransformMakeTranslation(centering_margin_x - move_x, centering_margin_y - move_y)];

		// fill path
		if([fillColor isKindOfClass:[UIColor class]])
		{
			[((UIColor *)fillColor) setFill];
			[path fill];
		}

		// stroke path
		if([strokeColor isKindOfClass:[UIColor class]])
		{
			[((UIColor *)strokeColor) setStroke];
			[path stroke];
		}
	}
}

@end

File: Classes/DrawUIBezierPathFitToUIView.swift
//
//  DrawUIBezierPathFitToUIView.swift
//
//  Created by Mossuru777 on 2018/01/23.
//  Copyright (c) 2018 Mossan. All rights reserved.
//

import UIKit
import CoreGraphics

public class DrawUIBezierPathFitToUIView {
    /// UIBezierPaths scale when fit to UIView.
    /// The value must be greater than 0.0 and 1.0 or less.
    public var scale: CGFloat {
        get {
            return self.scale
        }

        set {
            if newValue <= 0.0 || newValue > 1.0 {
                fatalError("ValueError: scale must be greater than 0.0 and 1.0 or less.")
            }
            self.scale = newValue
        }
    }

    private typealias PathInfo = (path: UIBezierPath, fillColor: UIColor?, strokeColor: UIColor?)

    private let viewSize: CGSize
    private var paths: [PathInfo]

    /// Initialize with UIView to fit UIBezierPaths.
    ///
    /// - Parameters:
    ///   - view: UIView to fit UIBezierPaths.
    ///   - scale: draw scale (default value: 1.0)
    public init(_ view: UIView, scale: CGFloat = 1.0) {
        self.viewSize = view.frame.size
        self.paths = []
        self.scale = scale
    }

    /// Adding fit target UIBezierPath.
    /// It is necessary to set a value for fillColor or strokeColor.
    ///
    /// - Parameters:
    ///   - path: Fit target UIBezierPath object.
    ///   - fillColor: Path fill color
    ///   - strokeColor: Path stroke color
    public func addPath(_ path: UIBezierPath, fillColor: UIColor?, strokeColor: UIColor?) {
        if fillColor == nil && strokeColor == nil {
            fatalError("NoDrawPathError: arguments indicate no draw path.")
        }

        // check exist path in array
        for path_info: PathInfo in self.paths {
            if path_info.path == path {
                return
            }
        }

        // add path to array
        paths.append((path, fillColor, strokeColor))
    }

    /// Draw scaled UIBezierPaths for fit UIView.
    /// Scaled UIBezierPath keeps original aspect ratio.
    public func drawScaledPathFitToView() {
        var totalRect = CGRect(x: .greatestFiniteMagnitude, y: .greatestFiniteMagnitude, width: 0.0, height: 0.0)

        var fit_scale: CGFloat
        var centering_margin_x: CGFloat = 0.0
        var centering_margin_y: CGFloat = 0
        var move_x: CGFloat
        var move_y: CGFloat

        // calculate total bounds / origin
        for path_info: PathInfo in self.paths {
            let path = path_info.path

            // reset path scaling
            path.apply(.identity)

            // extends bounds / origin
            if path_info.strokeColor != nil {
                // width
                if path.bounds.origin.x + path.bounds.size.width + (path.lineWidth / 2) > totalRect.size.width {
                    totalRect.size.width = path.bounds.origin.x + path.bounds.size.width + (path.lineWidth / 2)
                }

                // x origin
                if path.bounds.origin.x - (path.lineWidth / 2) < totalRect.origin.x {
                    totalRect.origin.x = path.bounds.origin.x - (path.lineWidth / 2)
                }

                // height
                if path.bounds.origin.y + path.bounds.size.height + (path.lineWidth / 2) > totalRect.size.height {
                    totalRect.size.height = path.bounds.origin.y + path.bounds.size.height + (path.lineWidth / 2)
                }

                // y origin
                if path.bounds.origin.y - (path.lineWidth / 2) < totalRect.origin.y {
                    totalRect.origin.y = path.bounds.origin.y - (path.lineWidth / 2)
                }
            } else if path_info.fillColor != nil {
                // width
                if path.bounds.origin.x + path.bounds.size.width > totalRect.size.width {
                    totalRect.size.width = path.bounds.origin.x + path.bounds.size.width
                }

                // x origin
                if path.bounds.origin.x < totalRect.origin.x {
                    totalRect.origin.x = path.bounds.origin.x
                }

                // height
                if path.bounds.origin.y + path.bounds.size.height > totalRect.size.height {
                    totalRect.size.height = path.bounds.origin.y + path.bounds.size.height
                }

                // y origin
                if path.bounds.origin.y < totalRect.origin.y {
                    totalRect.origin.y = path.bounds.origin.y
                }
            }
        }

        // calculate scale / centering margin / move
        if self.scale > 1.0 || self.scale <= 0 {
            self.scale = 1.0
        }
        if self.viewSize.width / totalRect.size.width < self.viewSize.height / totalRect.size.height {
            fit_scale = self.viewSize.width / totalRect.size.width * self.scale
        } else {
            fit_scale = self.viewSize.height / totalRect.size.height * self.scale
        }
        centering_margin_x = (self.viewSize.width / 2) - (totalRect.size.width * fit_scale / 2)
        centering_margin_y = (self.viewSize.height / 2) - (totalRect.size.height * fit_scale / 2)
        move_x = totalRect.origin.x * fit_scale
        move_y = totalRect.origin.y * fit_scale

        // draw scaled path
        for path_info: PathInfo in self.paths {
            let path = path_info.path

            // scaling path
            path.lineWidth = path.lineWidth * fit_scale
            path.apply(CGAffineTransform(scaleX: fit_scale, y: fit_scale))

            // move path
            path.apply(CGAffineTransform(translationX: centering_margin_x - move_x, y: centering_margin_y - move_y))

            // fill path
            if let fillColor = path_info.fillColor {
                fillColor.setFill()
                path.fill()
            }

            // stroke path
            if let strokeColor = path_info.strokeColor {
                strokeColor.setStroke()
                path.stroke()
            }
        }
    }
}
